resolwe/observers/mixins.py

```diff
@@ -0,0 +1,63 @@
+"""Mixins for Observable ViewSets."""
+import json
+
+from django.contrib.auth.models import User
+from django.contrib.contenttypes.models import ContentType
+
+from rest_framework.decorators import action
+from rest_framework.exceptions import NotFound
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from .models import Subscription
+from .protocol import CHANGE_TYPE_CREATE, CHANGE_TYPE_DELETE, CHANGE_TYPE_UPDATE
+
+
+class ObservableMixin:
+    """A Mixin to make a model ViewSet observable.
+
+    Adds the /subscribe and /unsubscribe endpoints to the list view.
+    """
+
+    def user_has_permission(self, id: int, user: User) -> bool:
+        """Verify that an object exists for a given user."""
+        return self.get_queryset().filter(pk=id).filter_for_user(user).exists()
+
+    @action(detail=False, methods=["post"])
+    def subscribe(self, request: Request) -> Response:
+        """Register an Observer for a resource."""
+        ids = dict(request.query_params).get("ids", None)
```

Comment:
Getting values from a QueryParams object discards all but the first value. Example query:
```
?session_id=2&ids=1&ids=2&ids=3
```
produces
```
dict(query_params) -> {"session_id": [2], "ids": [1, 2, 3]}
query_params -> {"session_id": 2, "ids": 1}
```
In the instance of session_id, we're not expecting mutlitple values, so calling `get` on QueryParams is fine. 
This would also work for the second line:
`dict(request.query_params).get("session_id")[0]`
but it would throw a (more vague) IndexError in case session_id was missing.